### PR TITLE
[SQLite] Bump and use byte[] variants added to blob_open for performance

### PR DESCRIPTION
**Pull request by Therzok**
Fixes https://github.com/dotnet/roslyn/issues/23424

NOTE: Currently using CI as a build bot, roslyn builds take too much time on my machine.

<details><summary>Ask Mode template not completed</summary>

### Customer scenario

What does the customer do to get into this situation, and why do we think this
is common enough to address for this release.  (Granted, sometimes this will be
obvious "Open project, VS crashes" but in general, I need to understand how
common a scenario is)

### Bugs this fixes

(either VSO or GitHub links)

### Workarounds, if any

Also, why we think they are insufficient for RC vs. RC2, RC3, or RTW

### Risk

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

### Performance impact

(with a brief justification for that assessment (e.g. "Low perf impact because no extra allocations/no complexity changes" vs. "Low")

### Root cause analysis

How did we miss it?  What tests are we adding to guard against it in the future?

### How was the bug found?

(E.g. customer reported it vs. ad hoc testing)

### Test documentation updated?

If this is a new non-compiler feature or a significant improvement to an existing feature, update https://github.com/dotnet/roslyn/wiki/Manual-Testing once you know which release it is targeting.

</details>

### CODE REVIEW

**Comment by Therzok** [comment]
I went with the route of adding another property (`<Name>Ptr`) simply due to the fact that string versions may also be used in other places.

File: src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/SqlConnection.cs [hunk 1]
```diff
@@ -204,10 +212,12 @@ public Stream ReadBlob(string dataTableName, string dataColumnName, long rowId)
             return stream;
         }
 
-        private Stream ReadBlob_InTransaction(string tableName, string columnName, long rowId)
+
+        private static byte[] mainPtr = Util.to_utf8 ("main");
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/SqlConnection.cs]: there is Util.ToUtf8 and Util.to_utf8?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/SqlConnection.cs]: No space before (

File: src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/SqlConnection.cs [hunk 2]
```diff
@@ -185,6 +185,14 @@ public int LastInsertRowId()
             => (int)raw.sqlite3_last_insert_rowid(_handle);
 
         public Stream ReadBlob(string dataTableName, string dataColumnName, long rowId)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/SqlConnection.cs]: does this overload need to exist?  who is calling into ReadBlob with a string for the table name or column name?

Comment (Therzok) [reply src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/SqlConnection.cs]: Well, it was public API, so not sure if it's safe to completely remove it (although the class is internal).

Comment (Therzok) [reply src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/SqlConnection.cs]: I can remove this overload if we don't need it (it's currently not used anywhere)

File: src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage.Accessor.cs [hunk 3]
```diff
@@ -146,7 +147,7 @@ private Stream ReadBlob(SqlConnection connection, TDatabaseId dataId)
                     // data for a row in our system can change, the ID will always stay the
                     // same, and the data will always be valid for our ID.  So there is no
                     // safety issue here.
-                    return connection.ReadBlob(DataTableName, DataColumnName, rowId);
+                    return connection.ReadBlob(DataTableNamePtr, DataColumnName, rowId);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage.Accessor.cs]: shouldn't the ColumnName be a byte[] too?

File: src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/SqlConnection.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/SqlConnection.cs]: Yes.  Please remove the overload.  This type is completely internal to roslyn.  We do not have to worry about breaking changes here.

File: src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage.Accessor.cs (hunk 3 above)
Comment (Therzok) [reply src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage.Accessor.cs]: Yep, good catches so far. I haven't compiled locally, so this probably has issues even at compile time :p

File: src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/Util.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis.SQLite
+{
+    internal static class Util
+    {
+        internal static byte[] ToUtf8(string sourceText)
```
Comment (Therzok) [inline src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/Util.cs]: @CyrusNajmabadi had forgotten to commit this file.

File: src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/Util.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+            {
+                return null;
+            }
+
+            byte[] byteArray;
+            int nlen = Encoding.UTF8.GetByteCount(sourceText) + 1;
+
+            byteArray = new byte[nlen];
+            nlen = Encoding.UTF8.GetBytes(sourceText, 0, sourceText.Length, byteArray, 0);
+            byteArray[nlen] = 0;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/Util.cs]: this should be documented why you're doing this (i presume sqlite requires 0 terminated utf8?).

Comment (Therzok) [reply src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/Util.cs]: Yes.

Comment (Therzok) [reply src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/Util.cs]: This was shamelessly copy pasted from sqlitepcl.raw :p

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/Util.cs]: Just to be clear.  By 'documented', i meant: 'comment in the code' :)

Comment (Therzok) [reply src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/Util.cs]: Yes, I did add some comments.

File: src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/Util.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
+﻿using System;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/Util.cs]: needs copyright.

File: src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/Util.cs [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    internal static class Util
+    {
+        internal static byte[] ToUtf8(string text)
+        {
+            if (text == null)
+            {
+                return null;
+            }
+
+            // SQLite expects null terminated UTF8, so do conversion by allocating
```
Comment (sharwell) [inline src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/Util.cs]: 💭 This method could be reduced to the following if you want it simple:

```
return Encoding.UTF8.GetBytes(text + '\0');
```

Comment (Therzok) [reply src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/Util.cs]: `Encoding.UTF8.GetBytes(string)` internally calls the byte[] variant, https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/Text/Encoding.cs#L770-L781

So, that would incur:
1 string allocation '\0' -> "0"
1 string allocation for the string concat (text + "0")

I used this method for performance.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/Util.cs]: > I used this method for performance.

Relative impact of costs needs to be considered.  So, for example, your change helps replace an unending amount of string->byte[] allocations with a single fixed number of allocations (<10).  Improving allocation scenarios doesn't mean removing all allocations, just the costly ones :)

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/Util.cs]: that said, i'm fine with the code as is.  It's nice and clear and doesn't do anything that seems unnecessary.  Seems like a good solution to me.

Comment (Therzok) [reply src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/Util.cs]: Ah, I just realized that I changed this locally before I moved the things to happen in the static ctors. I can simplify the code.

File: src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/Util.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis.SQLite
+{
+    internal static class Util
+    {
+        internal static byte[] ToUtf8(string text)
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/Util.cs]: This should probably be called "ToUtf8WithNullTerminator" or something. Seeing initial use my thought was "why do we need a helper?"

File: src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/SqlConnection.cs [hunk 9]
```diff
@@ -204,10 +204,12 @@ public Stream ReadBlob(string dataTableName, string dataColumnName, long rowId)
             return stream;
         }
 
-        private Stream ReadBlob_InTransaction(string tableName, string columnName, long rowId)
+
+        private static byte[] mainPtr = Util.ToUtf8("main");
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/SqlConnection.cs]: readonly? Yeah, the array is still mutable but it's a small protection.

File: src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage.cs [hunk 10]
```diff
@@ -55,6 +55,7 @@ internal partial class SQLitePersistentStorage : AbstractPersistentStorage
         ///  -----------------------------------------------
         /// </summary>
         private const string SolutionDataTableName = "SolutionData" + Version;
+        private static byte[] SolutionDataTableNamePtr = Util.ToUtf8(SolutionDataTableName);
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage.cs]: readonly everything here.

**Comment by jasonmalinowski** [comment]
Also, @sharwell or @jinujoe this will probably require a bump in the open source tool to track the new dependency. Who wants to own following up on that?

File: src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/SqlConnection.cs (hunk 9 above)
Comment (Therzok) [reply src/Workspaces/Core/Desktop/Workspace/SQLite/Interop/SqlConnection.cs]: Sure thing!

**Comment by heejaechang** [comment]
nice.

**Comment by Therzok** [comment]
Should be ok now.

**Comment by Therzok** [comment]
I think 15-7 is what we want.

**Comment by Therzok** [comment]
Fixed the Windows VisualStudio projects. Build should go green soon.

Opened another issue about other projects not using the Packages.props values. https://github.com/dotnet/roslyn/issues/25148

**Comment by Therzok** [comment]
@jasonmalinowski any idea what I might be doing wrong with the nuget package version?

**Comment by jasonmalinowski** [comment]
It looks like the package changed from having assets under the win7-x86 RID it's now jus win-x86.

**Comment by Therzok** [comment]
Thanks! Fingers crossed!

**Comment by Therzok** [comment]
Rebased on top of current dev15.7.x

**Comment by Therzok** [comment]
Which branch should I target this against?

**Comment by Therzok** [comment]
Failures look unrelated?

**Comment by heejaechang** [comment]
@dotnet/roslyn-infrastructure integration test failure? is it known issue?

https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_debug_vs-integration_prtest/1548/testReport/junit/Roslyn.VisualStudio.IntegrationTests.CSharp/CSharpReplIdeFeatures/AddUsing/

**Comment by heejaechang** [comment]
@jinujoseph no idea on when we are supposed to move the version up.

**Comment by jinujoseph** [comment]
new version bump will need paper work and  understanding of why we would want to move .

**Comment by jasonmalinowski** [comment]
And I should comment: the paperwork is sometimes just rubber-stamped by a tool, so it's possibly quite easy.

**Comment by Therzok** [comment]
Fwiw, I saw submissions for 1.1.11 in osstool

**Comment by jasonmalinowski** [comment]
@Therzok Cool, then it might get rubber stamped. @jinujoseph who is doing the paperwork then, at least to see if it gets the rubber stamp?

**Comment by jinujoseph** [comment]
@ivanbasov is helping with the paperwork here..thanks

**Comment by ivanbasov** [comment]
I've sent a request for the OSS permission approval for sqlite. I will keep you informed.

**Comment by jasonmalinowski** [comment]
@jinujoseph This just needs an delegated M2 approval then.

**Comment by jinujoseph** [comment]
Approved to merge for 15.8.Preview3

**Pull request merged**